src/main/java/org/gradle/android/VersionNumber.java

```diff
@@ -0,0 +1,298 @@
+/*
+ * Copyright 2012 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.android;
+
+import com.google.common.base.Objects;
+import com.google.common.collect.Ordering;
+
+import javax.annotation.Nullable;
+
+/**
+ * Represents, parses, and compares version numbers. Supports a couple of different schemes: <ul> <li>MAJOR.MINOR.MICRO-QUALIFIER (the default).</li> <li>MAJOR.MINOR.MICRO.PATCH-QUALIFIER.</li> </ul>
+ *
+ * <p>The {@link #parse} method handles missing parts and allows "." to be used instead of "-", and "_" to be used instead of "." for the patch number.
+ *
+ * <p>This class considers missing parts to be 0, so that "1.0" == "1.0.0" == "1.0.0_0".</p>
+ *
+ * <p>Note that this class considers "1.2.3-something" less than "1.2.3". Qualifiers are compared lexicographically ("1.2.3-alpha" < "1.2.3-beta") and case-insensitive ("1.2.3-alpha" <
+ * "1.2.3.RELEASE").
+ *
+ * <p>To check if a version number is at least "1.2.3", disregarding a potential qualifier like "beta", use {@code version.getBaseVersion().compareTo(VersionNumber.parse("1.2.3")) >= 0}.
+ *
+ * <p>This implementation comes from Gradle 7.4.2.
+ */
+public class VersionNumber implements Comparable<VersionNumber> {
```

Comment:
why have we copied the implementation of this class?
in general its good to add a comment to explain this.

Reply:
This class is deprecated and scheduled for removal. The choice was between this or bringing in a 3rd party lib such as [zafarkhaja/jsemver](https://github.com/zafarkhaja/jsemver). Doing this seemed the least disruptive. But I'll add that to the javadoc here.

Reply:
Can we use the non-deprecated (but internal) `org.gradle.util.internal.VersionNumber` in order to get rid of the deprecation warnings in the Build Scan?

Reply:
I think I'd rather just have a copy of the class.  Using an internal class opens us up to the functionality changing at any moment without notice.  Admittedly, though, this functionality is unlikely to change.  Given the option between the two bad choices, I think I'd prefer to just have our own copy of the class, but I don't feel strongly about it, either.

Reply:
Also, I gave the internal package a shot and wound up with failures like [these](https://ge.solutions-team.gradle.com/s/louve5sez62ia/tests/:testAndroid7_0_4/org.gradle.android.CrossVersionOutcomeAndRelocationTest/simple%20Android%20app%20is%20relocatable%20with%20Gradle%207.0.2%20and%20Android%20plugin%207.0.4?top-execution=1).

Reply:
Ok, let's keep using the copy then.